### PR TITLE
Initial version of hierarchy_network_demo

### LINKED ISSUE #570: Create a Python network example with hierarchy
Get some examples with hierarchy and the old temporal pooling in as a starting place for the community.

Useful code examples:

- https://github.com/numenta/nupic/blob/master/examples/network/network_api_demo.py

### PR DESCRIPTION
**Pull request by rmccall84**
fixes #570

Also fixed an out-of-date path to the hot-gym data in the network_api_demo.py

### CODE REVIEW

**Comment by rhyolight** [comment]
Updated to `Fixes #570`.

**Comment by rhyolight** [comment]
test

**Comment by rhyolight** [comment]
test

**Comment by rhyolight** [comment]
Ignore the "Contributor Validator" on this PR. Ryan has signed the CL, he just didn't have his local `.gitconfig` set up to identify himself properly so it isn't associated with him. We've chatted about fixing it.

**Comment by rhyolight** [comment]
Network failure on Travis-CI job. I've restarted it.

**Comment by rmccall84** [comment]
@scottpurdy let's review this

File: examples/network/hierarchy_network_demo.py [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+                                          json.dumps(SP_PARAMS))
+  # Make sure learning is enabled
+  spatialPoolerRegion.setParameter("learningMode", True)
+  # We want temporal anomalies so disable anomalyMode in the SP. This mode is
+  # used for computing anomalies in a non-temporal model.
+  spatialPoolerRegion.setParameter("anomalyMode", False)
+  return spatialPoolerRegion
+
+
+def createTemporalPooler(network, name):
```
Comment (scottpurdy) [inline examples/network/hierarchy_network_demo.py]: I know the region is called TPRegion but please use "temporal memory" to refer to or name variables. We are moving away from calling it a temporal pooler since it isn't doing temporal pooling.

**Comment by scottpurdy** [comment]
Can you paste the output from running this?

File: nupic/encoders/multi.py [hunk 2]
```diff
@@ -44,6 +44,7 @@ class MultiEncoder(Encoder):
   """A MultiEncoder encodes a dictionary or object with
   multiple components. A MultiEncode contains a number
   of sub-encoders, each of which encodes a separate component."""
+  #TODO expand this docstring.
```
Comment (scottpurdy) [inline nupic/encoders/multi.py]: Space between `#` and start of comment:

```# TODO: Expand this docstring```

File: nupic/data/file_record_stream.py [hunk 3]
```diff
@@ -23,9 +23,9 @@
 CSV file based implementation of a record stream
 
 FileRecordStream is class that can read and write .csv files that contain
-records. The file has 3 header lines that contain for each field the name, type
-and a special indicator for the fields that serve as the reset field,
-sequence id field and timestamp field. The other fields
+records. The file has 3 header lines that contain, for each field, the name
+(line 1), type (line 2), and a special indicator (line 3) for those fields that
```
Comment (scottpurdy) [inline nupic/data/file_record_stream.py]: I'd end the sentence after `special indicator (line 3)`. And then start a new sentence: `The special field can indicate that the field specifies a reset, is a sequence ID, or is a timestamp for the record.`

File: examples/network/network_api_demo.py [hunk 4]
```diff
@@ -33,8 +33,8 @@
 
 _VERBOSITY = 0  # how chatty the demo should be
 _SEED = 1956  # the random seed used throughout
-_DATA_PATH = "extra/hotgym/rec-center-hourly.csv"
-_OUTPUT_PATH = "test_output.csv"
+_DATA_PATH = "../prediction/data/extra/hotgym/rec-center-hourly.csv"
```
Comment (scottpurdy) [inline examples/network/network_api_demo.py]: You should not have to do this. Please undo this change. If it doesn't work, you may need to set `NTA_DATA_PATH` but it should default correctly.

File: examples/network/hierarchy_network_demo.py [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+_L1_SPATIAL_POOLER = "l1SpatialPoolerRegion"
+_L1_TEMPORAL_POOLER = "l1TemporalPoolerRegion"
+_L1_CLASSIFIER = "l1Classifier"
+
+_L2_SPATIAL_POOLER = "l2SpatialPoolerRegion"
+_L2_TEMPORAL_POOLER = "l2TemporalPoolerRegion"
+_L2_CLASSIFIER = "l2Classifier"
+
+
+def createEncoder():
```
Comment (scottpurdy) [inline examples/network/hierarchy_network_demo.py]: Please put three blank lines before top level functions, classes, __main__ block (see Python code style guide).

**Comment by rmccall84** [comment]
@scottpurdy 
Console output:
`Creating network
Running network
Writing output to: /Users/rmccall/nta/nupic/examples/network/hierarchy-demo-output.csv
L1 ave abs class. error: 5.545824
L2 ave abs class. error: 6.724396
Hierarchy demo finished`

CSV file output: 
`0,21.200000762939453,1.0,1.0
1,16.399999618530273,1.0,1.0
2,4.699999809265137,1.0,1.0
3,4.699999809265137,1.0,1.0
4,4.599999904632568,1.0,1.0
5,23.5,1.0,1.0`
